### PR TITLE
Set/unset alarms (configurable)

### PR DESCRIPTION
**Pull request by odin568**
Hi @darkphnx 
I have one more feature I need.

As said I want to split with this project one calendar into multiple logical.
There I also need different alarms (as they will be imported to another tool which handles them).

As alarms are something total different than filters, I added a dedicated section.

Not sure if this feature is already going too far for an "ical-filter-proxy".

### CODE REVIEW

**Comment by odin568** [comment]
Hmm... good point. I just saw in the rfc document that also absolute datetime can be defined, so quite flexible. But having a fixed date in our use case attached to all events probably won't make sense.
So in fact we could rely on ISO8601 format. Not sure how to simplify it - format is already quite readable (for techies 😺 ).

So simplest solution would be to add link to documentation about format and to check the format by parsing it into time in order to check if it is technically valid entry

**Comment by odin568** [comment]
Or we just support i.e. "\d+ days" and "\d+ hours" simply....

**Comment by odin568** [comment]
Implemented something, please have a look @darkphnx

File: lib/ical_filter_proxy/calendar_builder.rb [hunk 1]
```diff
@@ -33,5 +34,37 @@ def add_rules
...
+      triggers = alarms["triggers"]
+      return unless triggers
+
+      triggers.each do |trigger|
+        calendar.add_trigger(generate_iso_format(trigger))
+        end
+    end
+
+    def generate_iso_format(input)
+      if /(\d+)\s+days?/i =~ input
```
Comment (darkphnx) [inline lib/ical_filter_proxy/calendar_builder.rb:54]: You could replace all if these if statements with a `case / else` statement.

File: lib/ical_filter_proxy/calendar.rb [hunk 2]
```diff
@@ -1,24 +1,47 @@
...
+    attr_accessor :ical_url, :api_key, :timezone, :filter_rules, :clear_existing_alarms, :triggers
 
     def initialize(ical_url, api_key, timezone = 'UTC')
       self.ical_url = ical_url
       self.api_key = api_key
       self.timezone = timezone
 
       self.filter_rules = []
+      self.clear_existing_alarms = false
+      self.triggers = []
```
Comment (darkphnx) [inline lib/ical_filter_proxy/calendar.rb:12]: Can we call this `alarm_triggers` to be clearer?

File: lib/ical_filter_proxy/calendar.rb [hunk 3]
```diff
@@ -1,24 +1,47 @@
...
     end
 
     def add_rule(field, operator, value)
       self.filter_rules << FilterRule.new(field, operator, value)
     end
 
+    def add_trigger(trigger)
+      self.triggers.append(trigger)
```
Comment (darkphnx) [inline lib/ical_filter_proxy/calendar.rb:20]: `self.triggers << trigger` would be more idiomatic ruby.

File: lib/ical_filter_proxy/calendar_builder.rb [hunk 4]
```diff
@@ -33,5 +34,37 @@ def add_rules
...
+
+      if alarms["clear_existing"] == true
+        calendar.set_clear_existing_alarms
+      end
+
+      triggers = alarms["triggers"]
+      return unless triggers
+
+      triggers.each do |trigger|
+        calendar.add_trigger(generate_iso_format(trigger))
```
Comment (darkphnx) [inline lib/ical_filter_proxy/calendar_builder.rb:49]: Can we get this down to be as simple as `calendar.add_rule` above?

We could go about this by creating a new `AlarmTrigger` class similar to `FilterRule` and move the `generate_iso_format` logic into that too.

File: lib/ical_filter_proxy/calendar.rb [hunk 5]
```diff
@@ -1,24 +1,47 @@
...
+    end
+
+    def set_clear_existing_alarms
```
Comment (darkphnx) [inline lib/ical_filter_proxy/calendar.rb:23]: Probably unnecessary, you could just do `calendar.clear_existing_alarms = true` in CalendarBuilder.

**Review by darkphnx — COMMENTED** [review]
I think a solid approach. A little refactoring to tidy up would be great.

File: lib/ical_filter_proxy/calendar_builder.rb (hunk 1 above)
Comment (odin568) [reply lib/ical_filter_proxy/calendar_builder.rb]: Done.

File: lib/ical_filter_proxy/calendar.rb (hunk 2 above)
Comment (odin568) [reply lib/ical_filter_proxy/calendar.rb]: Done.

File: lib/ical_filter_proxy/calendar.rb (hunk 3 above)
Comment (odin568) [reply lib/ical_filter_proxy/calendar.rb]: Done.

File: lib/ical_filter_proxy/calendar.rb (hunk 5 above)
Comment (odin568) [reply lib/ical_filter_proxy/calendar.rb]: Done

File: lib/ical_filter_proxy/calendar_builder.rb [hunk 6]
```diff
@@ -33,5 +34,36 @@ def add_rules
...
+
+      triggers.each do |trigger|
+        calendar.add_alarm_trigger(generate_iso_format(trigger))
+        end
+    end
+
+    def generate_iso_format(input)
+      case input
+      when /(\d+)\s+days?/i
+        return "-P#{$1}D"
```
Comment (darkphnx) [inline lib/ical_filter_proxy/calendar_builder.rb:56]: You don't need the explicit returns, ruby will bail out of the case statement here and it'll be returned implicitly

Comment (odin568) [reply lib/ical_filter_proxy/calendar_builder.rb]: allright, adjusted

File: lib/ical_filter_proxy/calendar_builder.rb (hunk 4 above)
Comment (odin568) [reply lib/ical_filter_proxy/calendar_builder.rb]: Will do. Funnily I had it like that but than reduced it to just the triggers, but now makes sense again ;)

Comment (odin568) [reply lib/ical_filter_proxy/calendar_builder.rb]: Added it. Can you have a look again @darkphnx ?

File: lib/ical_filter_proxy/alarm_trigger.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
+module IcalFilterProxy
+  class AlarmTrigger
+
+    attr_accessor :alarm_trigger
+
+    def initialize(alarm_trigger)
+      self.alarm_trigger = alarm_trigger
```
Comment (darkphnx) [inline lib/ical_filter_proxy/alarm_trigger.rb:7]: Why not compile with `generate_iso_format` in the initialiser? Saves us having an additional get_trigger method.

File: lib/ical_filter_proxy/calendar_builder.rb [hunk 8]
```diff
@@ -33,5 +34,21 @@ def add_rules
       end
     end
 
+    def add_alarms
+      alarms = calendar_config["alarms"]
+      return unless alarms
+
+      if alarms["clear_existing"] == true
```
Comment (darkphnx) [inline lib/ical_filter_proxy/calendar_builder.rb:41]: A more idiomatic statement here would be

```ruby
calendar.clear_existing_alarms = true if alarms['clear_existing']
```

Or you could even go to:

```ruby
calendar.clear_existing_alarms = !!alarms['clear_existing']
```

File: lib/ical_filter_proxy/calendar_builder.rb [hunk 9]
```diff
@@ -33,5 +34,21 @@ def add_rules
...
+        calendar.clear_existing_alarms = true
+      end
+
+      triggers = alarms["triggers"]
+      return unless triggers
+
+      triggers.each do |trigger|
+        calendar.add_alarm_trigger(trigger)
+        end
```
Comment (darkphnx) [inline lib/ical_filter_proxy/calendar_builder.rb:50]: Nitpick: too much indentation

**Review by darkphnx — COMMENTED** [review]
This is looking good, just onto the nitpicks now, and then some tests.

File: lib/ical_filter_proxy/alarm_trigger.rb (hunk 7 above)
Comment (odin568) [reply lib/ical_filter_proxy/alarm_trigger.rb]: Yes agree, done.

File: lib/ical_filter_proxy/calendar_builder.rb (hunk 8 above)
Comment (odin568) [reply lib/ical_filter_proxy/calendar_builder.rb]: As a non-ruby native developer, first option is more understandable for me.
Done.

File: lib/ical_filter_proxy/calendar_builder.rb (hunk 9 above)
Comment (odin568) [reply lib/ical_filter_proxy/calendar_builder.rb]: Done.

**Comment by odin568** [comment]
So then only tests open so far

**Comment by odin568** [comment]
Added tests, you can also have a look here @darkphnx

**Review by darkphnx — APPROVED** [review]
Beautiful!

**Pull request merged**